### PR TITLE
21017 allow to not use the std files under windows

### PR DESCRIPTION
**Pull request by VincentBlondeau**
https://pharo.fogbugz.com/f/cases/21017/Allow-to-not-use-the-std-files-under-windows

### CODE REVIEW

**Comment by VincentBlondeau** [comment]
Failure of the build

**Comment by VincentBlondeau** [comment]
Windows VMs seems to be offline. Should be relaunched when online again

File: src/Files/NoStdio.class.st
Comment (pavel-krivanek) [inline src/Files/NoStdio.class.st]: As far as I know we have two versions of Windows VM. The one with the suffix "Console"  should be able to use standard streams. So then this platform check will fail.

Comment (VincentBlondeau) [reply src/Files/NoStdio.class.st]: Right... But how can I know that is a console or not?

Comment (estebanlm) [reply src/Files/NoStdio.class.st]: you can't, that's an important part of the problem (and why this is not solved long before). Only real way to know is to open the stdout stream and try to use it. If it fails, is not a console :(

Comment (estebanlm) [reply src/Files/NoStdio.class.st]: back to this. 
how do we solve this? (checking by binary name seems a weak strategy). I guess best is to add a primitive that can answer this (shame, but I guess needed).

what do you think?

**Review by estebanlm — DISMISSED** [review]
we need to first solve this "console" stuff :(

**Comment by akgrant43** [comment]
> I guess best is to add a primitive that can answer this (shame, but I guess needed).

I'm not sure that this is a shame.  Asking a plugin what its capabilities are is a reasonable strategy.

**Comment by VincentBlondeau** [comment]
So, a new primitive is needed in for the VM to solve that issue.

File: src/Files/NoStdio.class.st
Comment (akgrant43) [reply src/Files/NoStdio.class.st]: Hi Vincent,
I'm missing what the goal of this is.

At the moment it appears that both FileStream and Stdio force stdio streams to be written to regular files on Windows - I'm not sure why.

If that code is removed, it is possible to write to the console with PharoConsole.exe using FileStream>>stdout.  When I tried it with Stdio>>stdout it didn't produce anything.  I haven't investigated further.

But...

With PharoConsole.exe:

FileStream stdout closed 
" false"


With Pharo.exe:

FileStream stdout closed
" true"

Does that help?

Cheers,
Alistair

**Comment by akgrant43** [comment]
Hi Vincent,

While the driver for the primitive is to check stdio, it can be used for any file descriptor, and as a general rule I think it is better that all FilePlugin primitives are defined by FilePluginPrims, which means that if the primitive is enhanced in some way, e.g. better error reporting, the image has a single definition which needs to be updated.

And talking of better error reporting :-), the primitive provides additional information, i,e,:


```
	<primitive: 'primitiveFileDescriptorType' module: 'FilePlugin' error: error>
	error = #'bad argument' ifTrue: [ 
		fdNum isInteger ifFalse: 
			[ ^self error: 'fileDescriptor must be an integer' ] ].
	^ self primitiveFailed 
```

Stdio>>stdioHandles should also be moved to FilePluginPrims, but that can be a separate PR.

Thanks!
Alistair

**Comment by guillep** [comment]
Be careful with introducing circular dependencies. Files should not depend on FileSystem, it should be the other way around. Also, File also is a good placeholder for primitives, FilePluginPrims being specific to filesystem

**Comment by akgrant43** [comment]
Hi Guille,

> Be careful with introducing circular dependencies. Files should not depend on FileSystem, it should be the other way around.

Can you clarify where the circular dependency is?  I guess this is due to the order in which the bootstrap works?

> File also is a good placeholder for primitives, FilePluginPrims being specific to filesystem

I sent another email on File :-)  If I'm correct above about the issue being with the bootstrap order, perhaps we could introduce a new class, FilesPrims, which could be the central object for primitives for the Files package?  (I'd really like to avoid having primitive definitions scattered throughout the image, it tends to duplicate code and lead to poor definitions, e.g. error handling, and makes it more difficult to update the plugins).

Thanks,
Alistair

**Comment by jecisc** [comment]
Circular dependencies are for example

PackageA depends on PackageB and PackageB depends on PackageA.

In that case, which package do you loads first?

**Comment by akgrant43** [comment]
> Circular dependencies are for example
> 
> PackageA depends on PackageB and PackageB depends on PackageA.
> 
> In that case, which package do you loads first?

Either package.  And then leave it to Undeclared to take care of the issue. :-)

This may not work for the bootstrap though because Files and the FileSystem are central to basic operation.

**Comment by jecisc** [comment]
Currently dependencies are managed "by hand" in the bootstrap. But later I hope all dependencies will be declared in the baselines.

Metacello does not managed cycles in dependencies. In that case it might be a problem.

**Comment by VincentBlondeau** [comment]
A solution may to create another package with the File Primitives then? There are used in both Files and FileSsytem packages.

**Comment by akgrant43** [comment]
Hi Vincent,

> A solution may to create another package with the File Primitives then? There are used in both Files and FileSsytem packages.

If it really is the situation that Files is loaded before FileSystem (I'm going through the bootstrap process at the moment), then I'd prefer to move FilePluginPrims from FileSystem to Files and avoid the code duplication.

Also open to other ideas...

Cheers,
Alistair

**Comment by VincentBlondeau** [comment]
Hi Alistair,

I think we can move the primitive class to the Files package.  The class has only 3 dependencies: Collections-Native, Kernel and Zinc-Character-Encoding-Core. They are all already used by the Files package.

Cheers,
Vincent

**Comment by guillep** [comment]
It's exactly that. I don't want File to depend on FileSystem. It's already the other way around.

Also, why putting the primitives in a "FilePrims" placeholder class? And not in the File class directly? Let's think object oriented there :). Also putting the primitives in a separate class forces us to declare a method with an extra argument which is AFAIK less convenient for the JIT.

**Comment by guillep** [comment]
Actually the circular dependency is what made the build fail.

https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/Test%20pending%20pull%20request%20and%20branch%20Pipeline/detail/PR-694/17/pipeline/10

It probably entered into an infinite loop and failed due a timeout...

**Comment by akgrant43** [comment]
Hi Guille,

> Also, why putting the primitives in a "FilePrims" placeholder class?

This wasn't really a good idea.  As I suggested afterwards, I think we should move FilePluginPrims to the Files package - it is core to all file operations and should be one of the first things loaded.


> Let's think object oriented there :).

Right, FilePluginPrims is the image's view of the primitives - it is responsible for interpreting status values returned by the primitives, platform dependent argument conversion, any initialisation of the primitives that may be required, etc. :-)

If a set of primitives is only used by one class there is a good argument for defining the primitives in the class.  However we're not at that point yet in the streams tidy-up.


> Also putting the primitives in a separate class forces us to declare a method with an extra argument which is AFAIK less convenient for the JIT.

I don't understand this one at all, sorry.  Moving the primitive method from Stdio (or any other class) to FilePluginPrims doesn't add an argument.  (Just to be clear, I'm not questioning your statement about JIT optimisations).


> Actually the circular dependency is what made the build fail.

Hmm, Vincent's latest changes should have removed the dependency of Files on FileSystem.  If Vincent doesn't figure it out first I'll have a look (but don't have time today).

(It would still be great to be able to inspect .hermes files :-))

Cheers,
Alistair

**Comment by VincentBlondeau** [comment]
> Also putting the primitives in a separate class forces us to declare a method with an extra argument which is AFAIK less convenient for the JIT.

@guillep : That does work for the FFI calls but not for the primitives... You have to give the handle as an argument


> If a set of primitives is only used by one class there is a good argument for defining the primitives in the class. However we're not at that point yet in the streams tidy-up.

@akgrant43: the primitives are used by several classes (AbstractBinaryFileStream, File). So adding them to  the file class make them less easy to be used AFAIAA.

@guillep: I will put the primitives to the class side of the File class.

**Comment by guillep** [comment]
Hi,

I prefer to have the primitives in `File`'s class side :)

>>Also putting the primitives in a separate class forces us to declare a method with an extra >>argument which is AFAIK less convenient for the JIT.

>I don't understand this one at all, sorry. Moving the primitive method from Stdio (or any other >class) to FilePluginPrims doesn't add an argument. (Just to be clear, I'm not questioning your >statement about JIT optimisations).

My bad. Actually, as you may know, all primitives receive implicitly the receiver as first argument.
So, we can trick the interpreter by creating a method with an extra argument, and then the primitive will take that first argument as the receiver. This is used for example in `MirrorPrimitives`. Now, I was wrongly thinking that the file primitives receive a file as argument, and no, they receive a handle, so the receiver never matters :) So what I said before makes no sense for this case ;)

@VincentBlondeau I'll check later this afternoon why is this failing. Do you need help?
@akgrant43 About the hermes files, have you seen @tesonep's answer in the mailing list?

**Comment by akgrant43** [comment]
Hi Guille,

> I prefer to have the primitives in File's class side :)

I prefer to get rid of File altogether. :-)  As far as I can tell, File is just a bunch of utility methods with a completely misleading class comment.  But I haven't looked in to it properly, and don't want to hold up this PR, so I'm happy to go along with it for now.

> @akgrant43 About the hermes files, have you seen @tesonep's answer in the mailing list?

Yes, thanks.  My Pharo time is very limited at the moment, so I haven't had a chance to try it out, but will as part of continuing to look at this PR, assuming your or Vincent don't do that first (but it could take me quite a while).

Thanks!
Alistair

**Comment by guillep** [comment]
>I prefer to get rid of File altogether. :-) As far as I can tell, File is just a bunch of utility methods with a completely misleading class comment. But I haven't looked in to it properly, and don't want to hold up this PR, so I'm happy to go along with it for now.

Why you say so? I wrote that from scratch. It's exactly what you want from `FilePluginPrims`: a low level interface to files. However, it is built in an object oriented way, meant to encapsulate the access to the fileHandle, that users should not care about. Of course it is simple and small. But it is also composable :). People that don't want to drag the entire FileSystem dependency with them may use this simpler API.

From my point of view, the temporary class there is FilePluginPrims which is for sure there because at the moment File was not available and the old file streams suck.

I'll try to check with @VincentBlondeau later this afternoon

**Comment by VincentBlondeau** [comment]
Changes done:
- move the primitives to File
- get rid of FilePluginPrims
- deprecate File>>#imageFile

Waiting for CI to build

File: src/Files/File.class.st
Comment (guillep) [inline src/Files/File.class.st]: Why deprecating it? It was never a public method, and even less a method of this class. Nobody will call it. We can just remove it.

File: src/Files/File.class.st
Comment (guillep) [inline src/Files/File.class.st]: Do we need this primitive? It has no senders...

**Review by guillep — APPROVED** [review]
I still don't like the fact that now we are doing everywhere:

`File bla: handle`

`File bleh: handle`

To me that's a smell. I'd separate:
 - primitives that deal with files in File 
 - primitives that deal with streams in Stream/BinaryFileStream
 - primitives that deal with directories/filesystem into a Directory class

Like that, in each class we could do at worst `self class xxx` instead of hardcoding the `File` class, and `self xxx` at best (being xxx defined as `self class xxx` but leaving a hook for potential subclasses). That's how most primitives are in the image, I don't see why we want to split them into a separate class. Why not doing then an `IntegerPrimitives` or a `StringPrimitives` class?

Anyways, we can clean it in a second step in another PR.

**Comment by VincentBlondeau** [comment]
It will be in another PR ;) 
I removed the 2 methods like you ask

**Comment by VincentBlondeau** [comment]
Strange behavior on Unix CIs:

> + ./pharo Pharo.image test --junit-xml-output --stage-name=Tests-osx-32 '.*'
> 2018-05-14 23:26:44.714 Pharo[60038:507] ApplePersistenceIgnoreState: Existing state will not be touched. New state will be written to /var/folders/w0/nn_htp7n6ddcydvkv21hjzhc0000gn/T/org.pharo.Pharo.savedState
> 2018-05-14 23:26:45.076 Pharo[60038:507] 23:26:45.076 ERROR:     [0xa146e1a8] AQMEIO.cpp:377: _FindIOUnit: error -66680
> 2018-05-14 23:26:45.077 Pharo[60038:507] 23:26:45.077 ERROR:     [0xa146e1a8] >aq> AudioQueueObject.cpp:1590: Prime: failed (-66680); will stop (5512/0 frames)
> 2018-05-14 23:26:45.078 Pharo[60038:507] 23:26:45.077 ERROR:     [0xa146e1a8] AQMEIO.cpp:377: _FindIOUnit: error -66680
> 2018-05-14 23:26:45.078 Pharo[60038:507] 23:26:45.078 ERROR:     [0xa146e1a8] >aq> AudioQueueObject.cpp:1590: Prime: failed (-66680); will stop (33075/0 frames)
> 221'Cannot read stored value of #notExistingClassStoredSettingSTONString. Exception: KeyNotFound: key #MockNotExistingClass not found in SystemDictionary'
> \\somethingsomething./pharo: line 11: 60038 Abort trap: 6           "$DIR"/"pharo-vm/Pharo.app/Contents/MacOS/Pharo" --headless "$@"
> script returned exit code 134

File: src/FileSystem-Disk/DiskStore.class.st
Comment (dionisiydk) [inline src/FileSystem-Disk/DiskStore.class.st]: I think it would be better to name encoding methods more explicitly like:
- encodePathString:
- decodePathString:

**Review by dionisiydk — CHANGES_REQUESTED** [review]

**Comment by dionisiydk** [comment]
I agree with @guillep  it really smells.

I would expect that instead of "File bleh: handler" it would be "file bleh" where file variable will replace handler with real object instead of integer.

But now File is just file name without handler. And FileHandler is not part of Files package. 
So we will need to decide how to improve it.

**Comment by VincentBlondeau** [comment]
What is the difference between `FilePathEncoder encode: string`. and `File encode: string`?
Which one should be used if it is the same?

**Comment by VincentBlondeau** [comment]
@dionisiydk Yes. But in this case, one need to rewrite the primitives to use the instance variable of the receiver instead of an argument

**Comment by akgrant43** [comment]
Hi Guille,

> Why you say so? I wrote that from scratch.

Oh dear, put my foot in my mouth again, have I?  Please let me try to 
explain where I'm coming from:

The class comment for File starts with:

--
I represent a sequential binary File. I provide the minimum operations to:

- move the cursor fo the file
- reading
- writing

...
--

And then goes on to mention several methods that aren't actually
implemented, e.g. seekAbsolute:, seekRelative:, nextPutAll:, next:, etc.

This looks to me like something that has evolved over time and changed 
from its original goals.



> It's exactly what you want from FilePluginPrims: a low level interface 
> to files.

But FilePluginPrims isn't really a low level interface to files in the
sense of something that a normal programmer will use - it is
the in-image representation of FilePlugin, a collection of functions
needed to work with the OS filesystem.

The programmers interface to the file system is distributed across
FileReference, DiskDirectoryEntry and AbstractBinaryFileStream (and 
probably a couple of others that I've missed).  File picks out a few of 
these functions and makes them available separately.

The advantage of centralising the primitive methods in FilePluginPrims
is that it avoids duplicating these methods, and provides a central
place for some other functionality like string encoding / decoding that
the plugin expects.  (Having said that, I think I'm agreeing with you
that the primitives should be moved to the class that actually uses
them)


> However, it is built in an object oriented way, meant to encapsulate 
> the access to the fileHandle, that users should not care about.

While I'm sure that the classes that make up FileSystem aren't perfectly 
encapsulated, to suggest that they aren't object oriented is a bit 
unfair IMHO.



> Of 
> course it is simple and small. But it is also composable :). People 
> that don't want to drag the entire FileSystem dependency with them may 
> use this simpler API.

Given the bootstrap process that we are adopting, this is quite 
plausible / sensible.  But...

I think that File should have very well defined boundaries, otherwise it 
is just going to grow in to another FileSystem over time.  That's 
something that the class comments are missing at the moment.


> From my point of view, the temporary class there is FilePluginPrims 
> which is for sure there because at the moment File was not available 
> and the old file streams suck.

I don't know the history of FilePluginPrims, but it can be
refactored in to FileSystem without the need for File.


> To me that's a smell. I'd separate: 
> 
> primitives that deal with files in File 

If File is supposed to be a light-weight way of accessing the OS file 
system I don't see how it can't be something that is used by FileSystem 
(since by definition it doesn't include all the required functionality).


> primitives that deal with streams in Stream/BinaryFileStream 

AbstractBinaryFileStream.  The primitives are used by both 
BinaryFileStream and StdioStream.


> primitives that deal with directories/filesystem into a Directory class 

DiskDirectoryEntry

There's a couple other primitives that need to be moved as well, e.g. 
iterating over all entries in a directory.


> Like that, in each class we could do at worst self class xxx instead 
> of hardcoding the File class, and self xxx at best (being xxx defined 
> as self class xxx but leaving a hook for potential subclasses). That's 
> how most primitives are in the image, I don't see why we want to split 
> them into a separate class. Why not doing then an IntegerPrimitives or 
> a StringPrimitives class? 

> 
> Anyways, we can clean it in a second step in another PR. 

+1.


Cheers,
Alistair

**Comment by akgrant43** [comment]
>  But in this case, one need to rewrite the primitives to use the instance variable of the receiver instead of an argument

We want to be very careful before doing this.

The primitives access the instance variable by index number (AFAIK), so changing the class definition can require the primitive to be updated, i.e. it's a fairly brittle solution.

This is further exasperated by the fact that the primitives can be used by multiple smalltalks, e.g. Pharo, Squeak and NewSpeak, and referencing instance variables introduces the need to coordinate the definition across each flavour of smalltalk.

Cheers,
Alistair

**Comment by guillep** [comment]
Yeap, I'm with @akgrant43 there. I wouldn't try to rewrite the primitives, we have lived with them for already quite some time. As soon as it is encapsulated somewhere, thats ok to me.

**Comment by dionisiydk** [comment]
Yes. My sentence was not about primitives rewrite.
It just looks like we need object instead of integer to represent handle in file streams. Then file streams will just ask handle to do operations instead of ask some third man (File) to do operation with handle.
Of course on the low level of this FileHandle we will still have same primitives. But they will be hidden from users

**Comment by VincentBlondeau** [comment]
Hum... :
./bootstrap/scripts/3-prepare.sh: line 10: ./pharo: No such file or directory

**Comment by VincentBlondeau** [comment]
2018-05-21 19:50:00 (19.6 MB/s) - written to stdout [5829/5829]
Downloading the latest pharoVM:
	http://files.pharo.org/get-files/61/pharo-linux-stable.zip
script returned exit code 56

**Comment by guillep** [comment]
@VincentBlondeau please, don't do a new commit to relaunch the CI. You can connect to the CI and press the button *relaunch* :)

I'm rebuilding it now. As soon as it is green, I'll merge it.

@akgrant43 to answer some of your questions:

>The class comment for File starts with: [...]

Ok, so I should rework the comment :)
Just that you know, and for the record, File & co were originally designed to replace MultiByteOrTextAndMaybeBufferedStream & co. BinaryFileStream and File are two classes because one is to access the stream to an open file and the other to operate in the file without necessarily an open handle to it.

> The advantage of centralising the primitive methods in FilePluginPrims
is that it avoids duplicating these methods, and provides a central
place for some other functionality like string encoding / decoding that
the plugin expects. (Having said that, I think I'm agreeing with you
that the primitives should be moved to the class that actually uses
them)

To me the cleanest move is to make File & co the centralized way to access the promitives. 

> While I'm sure that the classes that make up FileSystem aren't perfectly
encapsulated, to suggest that they aren't object oriented is a bit
unfair IMHO.

Not talking about FileSystem in general, FileSystem is a pretty good library, but it found the same problems we did: MultiByteEverythingStream. And it tried to tackle it... What's not object oriented is the approach of the FileSystemPrimitives class to my understanding.

> I think that File should have very well defined boundaries, otherwise it
is just going to grow in to another FileSystem over time. That's
something that the class comments are missing at the moment.

This I agree. Most users should just use FileSystem to access the file system. Moreover, FileSystem abstracts you from the underlying storage. The File package only provides access to the file system in disk. Anyways, I want to keep the File package for internal usages of pharo (sources files, reading Hermes files... things for which loading FileSystem in the bootstrap is sometimes an overkill.

**Comment by guillep** [comment]
Test failures do not seem related

**Pull request merged**
